docs/source/plotly_directive.py

```diff
@@ -0,0 +1,151 @@
+"""
+Usage:
+```
+.. plotly::
+
+    import plotly.graph_objects as go
+    import numpy as np
+
+    x = np.linspace(0, 10, 100)
+    y = np.sin(x)
+
+    go.Figure(data=go.Scatter(x=t, y=y, mode='markers'))
+```
+
+Requirements:
+1. docstring contains a single code block.
+2. the code block ends with an expression that evaluates to a plotly figure.
+
+Based on:
+https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/sphinxext/plot_directive.py
+"""
+
+import os
+import textwrap
+import traceback
+
+from docutils.parsers.rst import Directive
+import jinja2  # Sphinx dependency.
+
+import plotly
+
+
+class PlotlyDirective(Directive):
+    """The ``.. plotly::`` directive, as documented in the module's docstring."""
+
+    has_content = True
+
+    def run(self):
+        """Run the plotly directive."""
+        try:
+            return run(
+                self.arguments,
+                self.content,
+                self.options,
+                self.state_machine,
+                self.state,
+                self.lineno,
+            )
+        except Exception as e:
+            raise self.error(str(e))
+
+
+def setup(app):
+    setup.app = app
+    setup.config = app.config
+    setup.confdir = app.confdir
+    app.add_directive("plotly", PlotlyDirective)
+
+    metadata = {
+        "parallel_read_safe": True,
+        "parallel_write_safe": True,
+        "version": plotly.__version__,
+    }
+    return metadata
+
+
+RST_TEMPLATE = """
+.. code-block:: python
+
+{{ source_code }}
+
+.. raw:: html
+
+   <iframe src="./{{ figure_name }}"
+     width="100%" height="500px" frameborder="0">
+   </iframe>
+"""
+
+
+class PlotError(RuntimeError):
+    pass
+
+
+def out_of_date(original, derived):
+    return os.stat(derived).st_mtime < os.stat(original).st_mtime
+
+
+def save_plotly_figure(fig, path):
+    fig_html = plotly.offline.plot(fig, output_type="div", include_plotlyjs="cdn", auto_open=False)
+    with open(path, "w") as f:
+        f.write(fig_html)
+
+
+def assign_last_line_into_variable(code, variable_name):
+    *rest, last = code.strip().split("\n")
+    last = "{} = ".format(variable_name) + last
+    return "\n".join([*rest, last])
+
+
+def run_code(code):
+    namespace = {}
+    variable_name = "fig"
+
+    try:
+        exec(assign_last_line_into_variable(code, variable_name), namespace)
+    except (Exception, SystemExit) as err:
+        raise PlotError(traceback.format_exc()) from err
+
+    return namespace[variable_name]
```

Comment:
## How to extract plotly figure from example code block:

Let's say we have code like this:
```
...
func_that_returns_plotly_figure()
```

`assign_last_line_into_variable` injects `fig = ` in the last line:

```
...
fig = func_that_returns_plotly_figure() 
```

Running this code generates a variable named `fig` which stores a plotly figure object.